// Copyright (c) 2026 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

package com.daml.ledger.api.v2;

import "com/daml/ledger/api/v2/event.proto";
import "com/daml/ledger/api/v2/reassignment.proto";
import "com/daml/ledger/api/v2/transaction_filter.proto";

option csharp_namespace = "Com.Daml.Ledger.Api.V2";
option java_outer_classname = "StateServiceOuterClass";
option java_package = "com.daml.ledger.api.v2";

// Allows clients to get state from the ledger.
service StateService {
  // Returns a stream of the snapshot of the active contracts and incomplete (un)assignments at a ledger offset.
  // Once the stream of GetActiveContractsResponses completes,
  // the client SHOULD begin streaming updates from the update service,
  // starting at the GetActiveContractsRequest.active_at_offset specified in this request.
  // Clients SHOULD NOT assume that the set of active contracts they receive reflects the state at the ledger end.
  rpc GetActiveContracts(GetActiveContractsRequest) returns (stream GetActiveContractsResponse);

  // Get the list of connected synchronizers at the time of the query.
  rpc GetConnectedSynchronizers(GetConnectedSynchronizersRequest) returns (GetConnectedSynchronizersResponse);

  // Get the current ledger end.
  // Subscriptions started with the returned offset will serve events after this RPC was called.
  rpc GetLedgerEnd(GetLedgerEndRequest) returns (GetLedgerEndResponse);

  // Get the latest successfully pruned ledger offsets
  rpc GetLatestPrunedOffsets(GetLatestPrunedOffsetsRequest) returns (GetLatestPrunedOffsetsResponse);
}

// If the given offset is different than the ledger end, and there are (un)assignments in-flight at the given offset,
// the snapshot may fail with "FAILED_PRECONDITION/PARTICIPANT_PRUNED_DATA_ACCESSED".
// Note that it is ok to request acs snapshots for party migration with offsets other than ledger end, because party
// migration is not concerned with incomplete (un)assignments.
message GetActiveContractsRequest {
  reserved 1;
  reserved "filter";

  reserved 2;
  reserved "verbose";

  // The offset at which the snapshot of the active contracts will be computed.
  // Must be no greater than the current ledger end offset.
  // Must be greater than or equal to the last pruning offset.
  // Required, must be a valid absolute offset (positive integer) or ledger begin offset (zero).
  // If zero, the empty set will be returned.
  int64 active_at_offset = 3;

  // Format of the contract_entries in the result. In case of CreatedEvent the presentation will be of
  // TRANSACTION_SHAPE_ACS_DELTA.
  // Required
  EventFormat event_format = 4;
}

message GetActiveContractsResponse {
  // The workflow ID used in command submission which corresponds to the contract_entry. Only set if
  // the ``workflow_id`` for the command was set.
  // Must be a valid LedgerString (as described in ``value.proto``).
  // Optional
  string workflow_id = 1;

  // For a contract there could be multiple contract_entry-s in the entire snapshot. These together define
  // the state of one contract in the snapshot.
  // A contract_entry is included in the result, if and only if there is at least one stakeholder party of the contract
  // that is hosted on the synchronizer at the time of the event and the party satisfies the
  // ``TransactionFilter`` in the query.
  oneof contract_entry {
    // The contract is active on the assigned synchronizer, meaning: there was an activation event on the given synchronizer (
    // created, assigned), which is not followed by a deactivation event (archived, unassigned) on the same
    // synchronizer, until the active_at_offset.
    // Since activeness is defined as a per synchronizer concept, it is possible, that a contract is active on one
    // synchronizer, but already archived on another.
    // There will be one such message for each synchronizer the contract is active on.
    ActiveContract active_contract = 2;

    // Included iff the unassigned event was before or at the active_at_offset, but there was no corresponding
    // assigned event before or at the active_at_offset.
    IncompleteUnassigned incomplete_unassigned = 3;

    // Important: this message is not indicating that the contract is active on the target synchronizer!
    // Included iff the assigned event was before or at the active_at_offset, but there was no corresponding
    // unassigned event before or at the active_at_offset.
    IncompleteAssigned incomplete_assigned = 4;
  }
}

message ActiveContract {
  // Required
  // The event as it appeared in the context of its last update (i.e. daml transaction or
  // reassignment). In particular, the last offset, node_id pair is preserved.
  // The last update is the most recent update created or assigned this contract on synchronizer_id synchronizer.
  // The offset of the CreatedEvent might point to an already pruned update, therefore it cannot necessarily be used
  // for lookups.
  CreatedEvent created_event = 1;

  // A valid synchronizer id
  // Required
  string synchronizer_id = 2;

  // Each corresponding assigned and unassigned event has the same reassignment_counter. This strictly increases
  // with each unassign command for the same contract. Creation of the contract corresponds to reassignment_counter
  // equals zero.
  // This field will be the reassignment_counter of the latest observable activation event on this synchronizer, which is
  // before the active_at_offset.
  // Required
  uint64 reassignment_counter = 3;
}

message IncompleteUnassigned {
  // Required
  // The event as it appeared in the context of its last activation update (i.e. daml transaction or
  // reassignment). In particular, the last activation offset, node_id pair is preserved.
  // The last activation update is the most recent update created or assigned this contract on synchronizer_id synchronizer before
  // the unassigned_event.
  // The offset of the CreatedEvent might point to an already pruned update, therefore it cannot necessarily be used
  // for lookups.
  CreatedEvent created_event = 1;

  // Required
  UnassignedEvent unassigned_event = 2;
}

message IncompleteAssigned {
  // Required
  AssignedEvent assigned_event = 1;
}

message GetConnectedSynchronizersRequest {
  // The party of interest
  // Must be a valid PartyIdString (as described in ``value.proto``).
  // If empty, all synchronizers this node is connected to will be returned
  // Optional
  string party = 1;

  // The id of a participant whose mapping of a party to connected synchronizers is requested.
  // Must be a valid participant-id retrieved through a prior call to getParticipantId.
  // Defaults to the participant id of the host participant.
  // Optional
  string participant_id = 2;

  // The ID of the identity provider configured by ``Identity Provider Config``
  // Optional, if not set, assume the user is managed by the default identity provider.
  string identity_provider_id = 3;
}

message GetConnectedSynchronizersResponse {
  message ConnectedSynchronizer {
    // The alias of the synchronizer
    // Required
    string synchronizer_alias = 1;

    // The ID of the synchronizer
    // Required
    string synchronizer_id = 2;

    // The permission on the synchronizer
    // Set if a party was used in the request, otherwise unspecified.
    // Optional
    ParticipantPermission permission = 3;
  }
  repeated ConnectedSynchronizer connected_synchronizers = 1;
}

// Enum indicating the permission level that the participant has for the party
// whose connected synchronizers are being listed.
enum ParticipantPermission {
  PARTICIPANT_PERMISSION_UNSPECIFIED = 0;

  PARTICIPANT_PERMISSION_SUBMISSION = 1;

  // participant can only confirm transactions
  PARTICIPANT_PERMISSION_CONFIRMATION = 2;

  // participant can only observe transactions
  PARTICIPANT_PERMISSION_OBSERVATION = 3;
}

message GetLedgerEndRequest {}

message GetLedgerEndResponse {
  // It will always be a non-negative integer.
  // If zero, the participant view of the ledger is empty.
  // If positive, the absolute offset of the ledger as viewed by the participant.
  int64 offset = 1;
}

message GetLatestPrunedOffsetsRequest {
  // Empty for now, but may contain fields in the future.
}

message GetLatestPrunedOffsetsResponse {
  // It will always be a non-negative integer.
  // If positive, the absolute offset up to which the ledger has been pruned,
  // disregarding the state of all divulged contracts pruning.
  // If zero, the ledger has not been pruned yet.
  int64 participant_pruned_up_to_inclusive = 1;

  // It will always be a non-negative integer.
  // If positive, the absolute offset up to which all divulged events have been pruned on the ledger.
  // It can be at or before the ``participant_pruned_up_to_inclusive`` offset.
  // For more details about all divulged events pruning,
  // see ``PruneRequest.prune_all_divulged_contracts`` in ``participant_pruning_service.proto``.
  // If zero, the divulged events have not been pruned yet.
  int64 all_divulged_contracts_pruned_up_to_inclusive = 2;
}
